fix(bug): surround union alternatives with parens when they contain s…

alexeagle opened the pull request:
…paces

fixes #88

mprobst commented:
So I presume this is because operator precedence in TypeScript's union types between `|` and something else doesn't match. What exactly doesn't match? Looking for whitespace seems like a hack that will bite us sooner or later.

alexeagle commented:
The two examples I know of are
string | () => any
and
() => any []

we could search the string for the arrow operator rather than space, that
solves all the cases I know of.
Looking at the spec for Arrow Function
https://github.com/Microsoft/TypeScript/blob/master/doc/spec.md#4.11
they mention precedence briefly but I don't see a grammar definition for
all the operators and their precedence.

On Tue, Sep 8, 2015 at 4:30 PM Martin Probst <notifications@github.com>
wrote:

> So I presume this is because operator precedence in TypeScript's union
> types between | and something else doesn't match. What exactly doesn't
> match? Looking for whitespace seems like a hack that will bite us sooner or
> later.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/angular/closure-to-dts/pull/92#issuecomment-138730695>
> .
>

mprobst commented:
The grammar that also specifies operator precedence through the rule ordering is here:
https://github.com/Microsoft/TypeScript/blob/master/doc/spec.md#a-grammar

You should be able to follow down from the primary type expression to see what is legal and what needs parentheses. The `Type` rule itself seems to be of particular interest, note how it's branching into `UnionOrIntersectionType` and `FunctionType` at that level, and note what rules are allowed within a `UnionType`.

alexeagle commented:
okay, I studied the grammar for a while, and I don't think I can precisely map that spec into an emitter. But, I do understand now why the ArrayType and UnionType need a PrimaryType rather than a non-PrimaryType which includes FunctionType and ConstructorType. So I think this is correct.

I added some comments explaining where these grammar productions come from, and changed to search the nested emit for the arrow (FunctionType token) or the ampersand (IntersectionType token).

If you want to comb through the grammar more precisely, then I need some help with that.

mprobst commented on src/main/java/com/google/javascript/cl2dts/DeclarationGenerator.java:
```diff
@@ -560,6 +561,32 @@ private void visitTypeDeclaration(JSType type, boolean isVarArgs) {
     }
 
+    /**
+     * Add parentheses to turn a Type grammar production into a PrimaryType.
```
3rd party `Adds parentheses...`

mprobst commented on src/main/java/com/google/javascript/cl2dts/DeclarationGenerator.java:
```diff
@@ -560,6 +561,32 @@ private void visitTypeDeclaration(JSType type, boolean isVarArgs) {
+    private void visitTypeAsPrimary(JSType type) {
+      // These tokens indicate that the type is non-primary, because it is one of:
+      // UnionType
+      // IntersectionType
```
Not `IntersectionType`, right?

mprobst commented:
LGTMs with one docs/language nit.
